### PR TITLE
[form-builder] Add overridable part for array input functionality 

### PR DESCRIPTION
**Pull request by rexxars**
Sorry for the vague description.

Basically, we have seen people wanting to customize the process of adding new items to an array, but once you go down that path you quickly come to a point where you might want to perform other operations, such as adding a button for sorting the array items alphabetically or similar.

This PR moves the "Add"-button logic into a separate part and makes both the primitive and object version of array inputs use it. I had to do a few generalization changes, but it does lead to more code reuse.

There is an example added in the tests studio where it allows you to add items by searching the list of possible types.

Have not had time to give the API as much thought as I would have wanted, open for suggestions, feel free to make changes directly in the PR.

### CODE REVIEW

File: packages/@sanity/form-builder/src/inputs/ArrayOfPrimitivesInput/ArrayOfPrimitivesInput.js [hunk 1]
```diff
@@ -212,17 +202,16 @@ export default class ArrayOfPrimitivesInput extends React.PureComponent<Props> {
...
-              )}
-            </div>
-          )}
+          <ArrayFunctions
+            type={type}
+            value={value}
+            readOnly={readOnly}
+            appendItem={this.append}
+            prependItem={this.prepend}
+            focusItem={this.handleFocusItem}
```
[the remark this answers is not in the document]
Comment (rexxars) [reply packages/@sanity/form-builder/src/inputs/ArrayOfPrimitivesInput/ArrayOfPrimitivesInput.js]: I'm a bit torn on this - `onFocusItem` sounds like an event handler to me. "When an item is focused, call this function", but this is the opposite - it's an action: "Focus the item with the given path for me". Same goes for the other items, `appendItem`, `focusItem` and `createValue` are also actions, not event handlers.

[the remark this answers is not in the document]
Comment (bjoerge) [reply packages/@sanity/form-builder/src/inputs/ArrayOfPrimitivesInput/ArrayOfPrimitivesInput.js]: I see your point. Personally, I like to think about it from the component's perspective, and let the component emit it blindly without having to consider context/what should happens next. When looking at it from the component's perspective it makes more sense to think of it as an event , e.g. "user requests the item to be focused/appended/whatever". So I think it should be up to up to the owner/consumer component to decide how to translate these events into the appropriate actions (or emit other events).

[the remark this answers is not in the document]
Comment (rexxars) [reply packages/@sanity/form-builder/src/inputs/ArrayOfPrimitivesInput/ArrayOfPrimitivesInput.js]: Will change for the sake of consistency.

**Comment by rexxars** [comment]
Updated to address review feedback, please take another look :)

**Pull request merged**